text/0000-model-lifecycle-hooks.md

```diff
@@ -0,0 +1,498 @@
+ *       }
+ *     }
+ *   ]
+ * }
+ */
+anotherPost.save()
+
+// Post#didUpdateData <RecordReference:post:2>
+// Post#didReceiveData <RecordReference:post:2>
+// Post#didUpdate <RecordReference:post:2>
+// Post#didUpdateRelationship: category, <BelongsToReference:category:1>
+// Post#didReceiveRelationship: category, <BelongsToReference:category:1>
+
+// ----------------------------------------------------------------------
+
+anotherPost.get("category");
+
+// Category#didInitData <RecordReference:category:1>
+// Category#didReceiveData <RecordReference:category:1>
+// Category#didUpdate <RecordReference:category:1>
+```
+
+
+## Use Cases
+
+### Check if there are comments for a post without triggering a request to server
+
+```js
+// app/models/post.js
+import Model from "ember-data/model";
+import { hasMany } from "ember-data/relationships";
+
+export default Model.extend({
+  comments: hasMany(),
+
+  didReceiveRelationshp(name, relationship) {
+    if (name === "comments") {
+      this.set("commentsLoaded", relationship.value() !== null);
+    }
+  }
+});
+```
+
+### Get all ids without triggering a request to server
+
+```js
+// app/models/post.js
+import Model from "ember-data/model";
+import { hasMany } from "ember-data/relationships";
+
+export default Model.extend({
+  comments: hasMany(),
+
+  didReceiveData(/* reference, payload */) {
+    this.set('commentIds', this.hasMany('comments').ids());
+  }
+});
+
+/**
+ * GET /posts/1
+ *
+ * data: {
+ *   id: 1,
+ *   type: 'post',
+ *   relationships: {
+ *     comments: {
+ *       data: [{ id: 1, type: 'comment' }]
+ *     }
+ *   }
+ * }
+ */
+store.findRecord('post', 1).then(function(post) {
+  assert.deepEqual(post.get("commentIds"), ["1"]);
+});
+```
+
+### Update properties based on the meta of a relationship
+
+```js
+// app/models/post.js
+import Model from "ember-data/model";
+import { hasMany } from "ember-data/relationships";
+
+export default Model.extend({
+  comments: hasMany(),
+
+  didReceiveData(/* reference, payload */) {
+    this.set('totalComments', this.hasMany('comments').meta().total);
+  }
+});
+
+/**
+ * GET /posts/1
+ *
+ * data: {
+ *   id: 1,
+ *   type: 'post',
+ *   relationships: {
+ *     comments: {
+ *       data: [{ id: 1, type: 'comment' }],
+ *       meta: { total: 123 }
+ *     }
+ *   }
+ * }
+ */
+store.findRecord('post', 1).then(function(post) {
+  assert.deepEqual(post.get("totalComments"), 123);
+});
+```
+
+# How We Teach This
+
+<!--
+What names and terminology work best for these concepts and why? How is this
+idea best presented? As a continuation of existing Ember patterns, or as a
+wholly new one?
+
+Would the acceptance of this proposal mean the Ember guides must be
+re-organized or altered? Does it change how Ember is taught to new users at any
+level?
+
+How should this feature be introduced and taught to existing Ember users?
+-->
+
+This enhances the known programming model of hooks by adding one for additional
```

Comment:
I don't think you mean "enhances"; perhaps you mean "leverages". In any case, this is unrelated to "How to Teach This". Note also that the mere fact that some programming model is known has nothing to do with whether it's the right approach for solving some particular problem.